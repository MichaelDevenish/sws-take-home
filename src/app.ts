
import express from 'express'
import {ApolloServer} from "@apollo/server";
import {expressMiddleware} from "@apollo/server/express4";
import bodyParser from "body-parser";
import cors from 'cors'
import {typeDefs, resolvers} from "./resolvers";
import {dataSources} from "./dataSources";

const app = express()

const server = new ApolloServer({ typeDefs, resolvers });

app.use(cors())
app.use(bodyParser.json())

async function startServer() {
    await server.start()
    app.use('/graphql', expressMiddleware(server, {
        context: async () => ({
            dataSources: dataSources()
        })
    }))
}

startServer();

export {
    app
}
