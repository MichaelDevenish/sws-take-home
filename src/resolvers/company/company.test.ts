import {app} from "../../app";
import request from "supertest";

describe("Test the company api", () => {
    test("It should list companies", async () => {
        const query =  `
          query test {
            listCompanies {
              id
              name
              tickerSymbol
              exchangeSymbol
              uniqueSymbol
              dateGenerated
              securityName
              exchangeCountryIso
              listingCurrencyIso
              canonicalUrl
              uniqueSymbolSlug
              scoreId
            }
          }
        `
        const response = await request(app)
            .post("/graphql")
            .send({query })
            .set('Accept', 'application/json')
            .set('Content-Type', 'application/json')

        expect(response.body).toMatchSnapshot()
        expect(response.status).toEqual(200)
    });

    test("It should be able to fetch a company with all relations", async () => {
        const query =  `
          query test {
            getCompany(companyId: "4BE2C01F-F390-479C-A166-8E0DD73CF7B9") {
              id
              name
              tickerSymbol
              exchangeSymbol
              uniqueSymbol
              dateGenerated
              securityName
              exchangeCountryIso
              listingCurrencyIso
              canonicalUrl
              uniqueSymbolSlug
              scoreId
              companyPriceClose {
                date
                companyId
                price
                dateCreated
              }
              companyPriceCloses {
                date
                companyId
                price
                dateCreated
              }
              companyScore {
                id
                companyId
                dateGenerated
                future
                health
                management
                past
                value
                misc
                total
                sentence
              }
            }
          }
        `
        const response = await request(app)
            .post("/graphql")
            .send({query })
            .set('Accept', 'application/json')
            .set('Content-Type', 'application/json')

        expect(response.body).toMatchSnapshot()
        expect(response.status).toEqual(200)
    });

    test("It should fail trying to load a campaign with no args", async () => {
        const query =  `
          query test {
            getCompany {
              id
            }
          }
        `
        const response = await request(app)
            .post("/graphql")
            .send({query })
            .set('Accept', 'application/json')
            .set('Content-Type', 'application/json')

        expect(response.body).toMatchSnapshot()
        expect(response.status).toEqual(400)
    });
});