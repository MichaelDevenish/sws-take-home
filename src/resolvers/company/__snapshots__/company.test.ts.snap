// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the company api It should be able to fetch a company with all relations 1`] = `
{
  "data": {
    "getCompany": {
      "canonicalUrl": "/stocks/au/materials/asx-bhp/bhp-group-shares",
      "companyPriceClose": {
        "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
        "date": "2020-05-22",
        "dateCreated": "2020-05-22 02:01:05.0566667",
        "price": 34.32,
      },
      "companyPriceCloses": [
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-22",
          "dateCreated": "2020-05-22 02:01:05.0566667",
          "price": 34.32,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-21",
          "dateCreated": "2020-05-21 02:01:08.3633333",
          "price": 34.51,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-20",
          "dateCreated": "2020-05-20 02:01:51.7800000",
          "price": 34.72,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-19",
          "dateCreated": "2020-05-19 02:01:14.0500000",
          "price": 35.05,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-18",
          "dateCreated": "2020-05-18 02:01:16.4200000",
          "price": 33.1,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-15",
          "dateCreated": "2020-05-15 02:02:20.8500000",
          "price": 31.67,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-14",
          "dateCreated": "2020-05-14 02:00:55.7433333",
          "price": 30.61,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-13",
          "dateCreated": "2020-05-13 02:00:52.0466667",
          "price": 31.01,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-12",
          "dateCreated": "2020-05-12 02:01:29.0300000",
          "price": 30.72,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-11",
          "dateCreated": "2020-05-11 02:01:51.7166667",
          "price": 31.55,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-08",
          "dateCreated": "2020-05-08 02:01:22.5800000",
          "price": 31.4,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-07",
          "dateCreated": "2020-05-07 02:01:10.7633333",
          "price": 31,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-06",
          "dateCreated": "2020-05-06 02:01:23.0800000",
          "price": 30.41,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-05",
          "dateCreated": "2020-05-05 02:01:12.2200000",
          "price": 30.69,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-04",
          "dateCreated": "2020-05-04 02:01:20.8200000",
          "price": 29.88,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-05-01",
          "dateCreated": "2020-05-01 02:00:52.1200000",
          "price": 29.84,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-30",
          "dateCreated": "2020-04-30 02:00:37.6933333",
          "price": 32.35,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-29",
          "dateCreated": "2020-04-29 02:00:43.5700000",
          "price": 31.12,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-28",
          "dateCreated": "2020-04-28 02:04:53.5933333",
          "price": 30.49,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-27",
          "dateCreated": "2020-04-27 02:04:35.8766667",
          "price": 30.5,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-24",
          "dateCreated": "2020-04-24 02:00:34.3666667",
          "price": 30.54,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-23",
          "dateCreated": "2020-04-23 02:00:42.3600000",
          "price": 29.75,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-22",
          "dateCreated": "2020-04-22 02:01:27.0300000",
          "price": 28.97,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-21",
          "dateCreated": "2020-04-21 02:01:21.4566667",
          "price": 30.05,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-20",
          "dateCreated": "2020-04-20 02:01:24.7433333",
          "price": 30.82,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-17",
          "dateCreated": "2020-04-17 02:01:21.8166667",
          "price": 31.28,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-16",
          "dateCreated": "2020-04-16 02:01:40.2300000",
          "price": 30.79,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-15",
          "dateCreated": "2020-04-15 02:01:38.6800000",
          "price": 31.43,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-14",
          "dateCreated": "2020-04-14 02:02:59.0000000",
          "price": 31.66,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-09",
          "dateCreated": "2020-04-09 02:01:35.0633333",
          "price": 31.5,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-08",
          "dateCreated": "2020-04-08 02:01:35.5200000",
          "price": 31.32,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-07",
          "dateCreated": "2020-04-07 02:01:25.2533333",
          "price": 31.7,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-06",
          "dateCreated": "2020-04-06 02:01:34.7433333",
          "price": 31.79,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-03",
          "dateCreated": "2020-04-03 01:01:34.0566667",
          "price": 30.33,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-02",
          "dateCreated": "2020-04-02 01:01:35.1933333",
          "price": 29.85,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-04-01",
          "dateCreated": "2020-04-01 01:01:30.5000000",
          "price": 30.23,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-03-31",
          "dateCreated": "2020-03-31 01:01:27.7966667",
          "price": 28.98,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-03-30",
          "dateCreated": "2020-03-30 01:03:47.7566667",
          "price": 30.2,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-03-27",
          "dateCreated": "2020-03-27 02:45:42.0833333",
          "price": 29.03,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-03-26",
          "dateCreated": "2020-03-26 01:01:29.3333333",
          "price": 30.67,
        },
        {
          "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
          "date": "2020-03-25",
          "dateCreated": "2020-03-25 01:01:33.0966667",
          "price": 31.3,
        },
      ],
      "companyScore": {
        "companyId": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
        "dateGenerated": "2020-05-24 10:37:17.000000",
        "future": 0,
        "health": 3,
        "id": "1895",
        "management": 0,
        "misc": 0,
        "past": 4,
        "sentence": "Proven track record average dividend payer.",
        "total": 12,
        "value": 1,
      },
      "dateGenerated": "2020-05-24 10:36:54.000000",
      "exchangeCountryIso": "AU",
      "exchangeSymbol": "ASX",
      "id": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
      "listingCurrencyIso": "AUD",
      "name": "BHP Group",
      "scoreId": 1895,
      "securityName": "LTD Ordinary Shares",
      "tickerSymbol": "BHP",
      "uniqueSymbol": "ASX:BHP",
      "uniqueSymbolSlug": "asx-bhp",
    },
  },
}
`;

exports[`Test the company api It should fail trying to load a campaign with no args 1`] = `
{
  "errors": [
    {
      "extensions": {
        "code": "GRAPHQL_VALIDATION_FAILED",
      },
      "locations": [
        {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "Field "getCompany" argument "companyId" of type "ID!" is required, but it was not provided.",
    },
  ],
}
`;

exports[`Test the company api It should list companies 1`] = `
{
  "data": {
    "listCompanies": [
      {
        "canonicalUrl": "/stocks/au/software/asx-apt/afterpay-shares",
        "dateGenerated": "2020-05-24 11:01:59.000000",
        "exchangeCountryIso": "AU",
        "exchangeSymbol": "ASX",
        "id": "46B285BC-B25F-4814-985C-390A4BFA2023",
        "listingCurrencyIso": "AUD",
        "name": "Afterpay",
        "scoreId": 26507,
        "securityName": "Ordinary Shares",
        "tickerSymbol": "APT",
        "uniqueSymbol": "ASX:APT",
        "uniqueSymbolSlug": "asx-apt",
      },
      {
        "canonicalUrl": "/stocks/au/materials/asx-bhp/bhp-group-shares",
        "dateGenerated": "2020-05-24 10:36:54.000000",
        "exchangeCountryIso": "AU",
        "exchangeSymbol": "ASX",
        "id": "4BE2C01F-F390-479C-A166-8E0DD73CF7B9",
        "listingCurrencyIso": "AUD",
        "name": "BHP Group",
        "scoreId": 1895,
        "securityName": "LTD Ordinary Shares",
        "tickerSymbol": "BHP",
        "uniqueSymbol": "ASX:BHP",
        "uniqueSymbolSlug": "asx-bhp",
      },
      {
        "canonicalUrl": "/stocks/us/retail/nyse-baba/alibaba-group-holding",
        "dateGenerated": "2020-05-24 06:23:25.000000",
        "exchangeCountryIso": "US",
        "exchangeSymbol": "NYSE",
        "id": "C8250615-877A-4AE1-BEF3-636B69CD83E8",
        "listingCurrencyIso": "USD",
        "name": "Alibaba Group Holding",
        "scoreId": 21152,
        "securityName": "ADS EACH REPR 1 ORD SHS (SPON) (CDI)",
        "tickerSymbol": "BABA",
        "uniqueSymbol": "NYSE:BABA",
        "uniqueSymbolSlug": "nyse-baba",
      },
      {
        "canonicalUrl": "/stocks/us/media/nasdaq-fb/facebook",
        "dateGenerated": "2020-05-24 04:51:37.000000",
        "exchangeCountryIso": "US",
        "exchangeSymbol": "NasdaqGS",
        "id": "8ACBC59A-0EC5-42D1-9F7B-6D31ABD4421E",
        "listingCurrencyIso": "USD",
        "name": "Facebook",
        "scoreId": 17129,
        "securityName": "Class A Common Stock",
        "tickerSymbol": "FB",
        "uniqueSymbol": "NasdaqGS:FB",
        "uniqueSymbolSlug": "nasdaq-fb",
      },
      {
        "canonicalUrl": "/stocks/us/software/nasdaq-msft/microsoft",
        "dateGenerated": "2020-05-24 03:39:14.000000",
        "exchangeCountryIso": "US",
        "exchangeSymbol": "NasdaqGS",
        "id": "424EB65E-8C34-42BF-A107-61F93D4E9E6D",
        "listingCurrencyIso": "USD",
        "name": "Microsoft",
        "scoreId": 81,
        "securityName": "Common Stock",
        "tickerSymbol": "MSFT",
        "uniqueSymbol": "NasdaqGS:MSFT",
        "uniqueSymbolSlug": "nasdaq-msft",
      },
      {
        "canonicalUrl": "/stocks/us/pharmaceuticals-biotech/nyse-pfe/pfizer",
        "dateGenerated": "2020-05-24 02:48:09.000000",
        "exchangeCountryIso": "US",
        "exchangeSymbol": "NYSE",
        "id": "3B07801A-1C3A-46E6-A7FF-E1F59F31C466",
        "listingCurrencyIso": "USD",
        "name": "Pfizer",
        "scoreId": 1471,
        "securityName": "Common Stock",
        "tickerSymbol": "PFE",
        "uniqueSymbol": "NYSE:PFE",
        "uniqueSymbolSlug": "nyse-pfe",
      },
      {
        "canonicalUrl": "/stocks/us/media/nyse-dis/walt-disney",
        "dateGenerated": "2020-05-24 02:40:27.000000",
        "exchangeCountryIso": "US",
        "exchangeSymbol": "NYSE",
        "id": "A0A1A293-FDA8-48DF-9EBF-35556CDE3235",
        "listingCurrencyIso": "USD",
        "name": "Walt Disney",
        "scoreId": 1629,
        "securityName": "Common Shares",
        "tickerSymbol": "DIS",
        "uniqueSymbol": "NYSE:DIS",
        "uniqueSymbolSlug": "nyse-dis",
      },
      {
        "canonicalUrl": "/stocks/us/retail/nasdaq-amzn/amazoncom",
        "dateGenerated": "2020-05-24 02:31:18.000000",
        "exchangeCountryIso": "US",
        "exchangeSymbol": "NasdaqGS",
        "id": "D0665877-9EC5-4568-8A29-E8FFF77DF072",
        "listingCurrencyIso": "USD",
        "name": "Amazon.com",
        "scoreId": 33,
        "securityName": "Common Stock",
        "tickerSymbol": "AMZ",
        "uniqueSymbol": "NasdaqGS:AMZ",
        "uniqueSymbolSlug": "nasdaq-amz",
      },
      {
        "canonicalUrl": "/stocks/us/tech/nasdaq-aapl/apple",
        "dateGenerated": "2020-05-24 02:20:41.000000",
        "exchangeCountryIso": "US",
        "exchangeSymbol": "NasdaqGS",
        "id": "743F0744-8987-4339-B565-DEE3A93E9934",
        "listingCurrencyIso": "USD",
        "name": "Apple",
        "scoreId": 162,
        "securityName": "Common Stock",
        "tickerSymbol": "AAPL",
        "uniqueSymbol": "NasdaqGS:AAPL",
        "uniqueSymbolSlug": "nasdaq-aapl",
      },
      {
        "canonicalUrl": "/stocks/au/telecom/asx-tls/telstra-shares",
        "dateGenerated": "2020-05-23 14:07:15.000000",
        "exchangeCountryIso": "AU",
        "exchangeSymbol": "ASX",
        "id": "FC7B296B-300B-4710-8F84-D68A5BFBC75B",
        "listingCurrencyIso": "AUD",
        "name": "Telstra",
        "scoreId": 1,
        "securityName": "Ordinary Shares",
        "tickerSymbol": "TLS",
        "uniqueSymbol": "ASX:TLS",
        "uniqueSymbolSlug": "asx-tls",
      },
      {
        "canonicalUrl": "/stocks/us/automobiles/nasdaq-tsla/tesla",
        "dateGenerated": "2020-05-23 08:32:49.000000",
        "exchangeCountryIso": "US",
        "exchangeSymbol": "NasdaqGS",
        "id": "65DCE5EA-70D6-417F-9CAC-1EAA92FB7F1C",
        "listingCurrencyIso": "USD",
        "name": "Tesla",
        "scoreId": 18906,
        "securityName": "Common Stock",
        "tickerSymbol": "TSLA",
        "uniqueSymbol": "NasdaqGS:TSLA",
        "uniqueSymbolSlug": "nasdaq-tsla",
      },
      {
        "canonicalUrl": "/stocks/us/transportation/nyse-dal/delta-air-lines",
        "dateGenerated": "2020-05-23 03:14:10.000000",
        "exchangeCountryIso": "US",
        "exchangeSymbol": "NYSE",
        "id": "49A0E7C9-F918-4E97-AECA-D8D37F9A3F4F",
        "listingCurrencyIso": "USD",
        "name": "Delta Air Lines",
        "scoreId": 2038,
        "securityName": "Common Stock",
        "tickerSymbol": "DAL",
        "uniqueSymbol": "NYSE:DAL",
        "uniqueSymbolSlug": "nyse-dal",
      },
    ],
  },
}
`;
